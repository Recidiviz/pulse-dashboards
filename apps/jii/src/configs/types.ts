// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { z } from "zod";

import { ParsedRecord, ResidentRecord } from "~datatypes";

import { EligibilityReport } from "../models/EligibilityReport/interface";
import { residentOpportunitySchemas } from "./residentsOpportunitySchemas";
import { stateCodes } from "./stateConstants";

export type ProfileField = {
  label: string;
  value: string;
  moreInfo?: string;
};

export type RequirementCopy = { criterion: string; ineligibleReason?: string };

export type OpportunityConfig = {
  urlSection: string;
  firestoreCollection: string;
  htmlTitle: string;
  copy: {
    headline: string;
    subheading: string;
    about: {
      title: string;
      body: string;
      linkText: string;
      fullPage: string;
      fullPageHtmlTitle: string;
    };
    requirements: {
      title: string;
      trackedCriteria: Record<string, RequirementCopy>;
      untrackedCriteria: Array<RequirementCopy>;
      staticRequirementsLabel: string;
      linkText: string;
      fullPage: string;
      fullPageHtmlTitle: string;
    };
    nextSteps: {
      title: string;
      body: string;
      linkText: string;
      fullPage: string;
      fullPageHtmlTitle: string;
    };
    menuLabel: string;
    formPreview: {
      title: string;
    };
  };
};

export type IncarcerationOpportunityId = "usMeSCCP";

export type ResidentsConfig = {
  headerProfileFields: Array<ProfileField>;
  incarcerationOpportunities: Partial<
    Record<IncarcerationOpportunityId, OpportunityConfig>
  >;
};

export type StateCode = z.infer<typeof stateCodes>;

export type ResidentOpportunitySchemaMapping =
  typeof residentOpportunitySchemas;

/**
 * Maps opportunity ID to its successfully parsed record format
 */
export type OpportunityRecord<O extends IncarcerationOpportunityId> =
  ParsedRecord<ResidentOpportunitySchemaMapping[O]>["output"];

/**
 * Mapping of opportunity IDs to their compatible EligibilityReport constructors
 */
export type ResidentEligibilityReportMapping = {
  [Id in IncarcerationOpportunityId]: new (
    ...args: [
      ResidentRecord["output"],
      OpportunityConfig,
      OpportunityRecord<Id> | undefined,
    ]
  ) => EligibilityReport;
};

export type StateConfig<Code extends StateCode = StateCode> = {
  stateCode: Code;
  displayName: string;
  urlSlug: string;
};

export type LandingPageConfig = {
  copy: {
    intro: string;
    selectorLabel: string;
    selectorPlaceholder: string;
  };
};

export type StateLandingPageConfig = {
  copy: {
    intro: string;
    selectorLabel: string;
    selectorPlaceholder: string;
    useCases: {
      intro: string;
      examples: Array<{
        icon: string;
        description: string;
      }>;
    };
  };
  connections: {
    [key: string]: { displayName: string; connectionName: string };
  };
};
