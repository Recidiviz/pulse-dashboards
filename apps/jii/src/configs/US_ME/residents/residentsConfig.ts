// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ResidentsConfig } from "../../types";
import { config } from "../usMeSCCP/config";
import releaseDateInfo from "./releaseDateInfo.md?raw";

export const usMeResidentsConfig: ResidentsConfig = {
  headerProfileFields: [
    {
      label: "Current release date",
      value:
        "{{#if resident.releaseDate}}{{formatFullDate resident.releaseDate}}{{else}}Not available{{/if}}",
      moreInfo: releaseDateInfo,
    },
  ],
  incarcerationOpportunities: {
    usMeSCCP: config,
  },
  home: {
    progress: {
      title: "Your Progress",
    },
    eligibility: {
      title: "Your Eligibility",
    },
    footer: {
      about: {
        title: "About this app",
        body: `This app is made by Recidiviz, a technology nonprofit.  
        It is free to access. The information updates daily.`,
      },
      contact: {
        title: "Contact us",
        body: `1322 Webster St, Suite 402  
        Oakland, CA 94612`,
      },
    },
  },
};
