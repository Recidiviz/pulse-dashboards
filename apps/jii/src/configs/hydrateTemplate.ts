// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { toTitleCase } from "@artsy/to-title-case";
import { isFuture } from "date-fns";
import Handlebars from "handlebars";

import { formatFullDate } from "../utils/date";

Handlebars.registerHelper("titleCase", (s: string) => toTitleCase(s));
Handlebars.registerHelper("lowerCase", (s: string) => s.toLowerCase());
Handlebars.registerHelper("formatFullDate", (d: Date) => formatFullDate(d));
Handlebars.registerHelper("isFutureDate", (d: Date) => isFuture(d));

/**
 * Hydrates the Handlebars template in `template` with the values in `context`.
 */
export function hydrateTemplate(
  template: string,
  context: Record<string, unknown>,
) {
  return Handlebars.compile(template)(context);
}
