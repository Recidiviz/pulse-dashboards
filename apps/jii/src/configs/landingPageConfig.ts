// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { LandingPageConfig } from "./types";

export const landingPageConfig: LandingPageConfig = {
  copy: {
    intro: `# Learn about opportunities you may be eligible for.`,
    selectorLabel: "Find opportunities in the state where you’re incarcerated",
    selectorPlaceholder: "Select a state …",
  },
  states: [{ stateCode: "US_ME", displayName: "Maine", urlSlug: "maine" }],
};
