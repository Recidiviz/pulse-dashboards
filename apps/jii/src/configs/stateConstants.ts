// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { keyBy } from "lodash";
import { z } from "zod";

import { StateCode, StateConfig } from "./types";

/**
 * This is the source of truth for valid, enabled state codes in the application
 */
export const stateCodes = z.enum(["US_ME"]);

/**
 * Given all state codes in tuple form, expects a corresponding tuple of config objects
 */
type StateConfigArray<StateCodeList> = StateCodeList extends [
  infer First extends StateCode,
  ...infer Rest,
]
  ? [StateConfig<First>, ...StateConfigArray<Rest>]
  : [];

/**
 * Exhaustive list of {@link StateConfig} objects for all supported states
 */
export const stateConfigs: StateConfigArray<typeof stateCodes.options> = [
  { stateCode: "US_ME", displayName: "Maine", urlSlug: "maine" },
];

/**
 * Exhaustive mapping of {@link StateConfig} objects for all supported states
 */
export const stateConfigsByStateCode: {
  [Code in StateCode]: StateConfig<Code>;
} =
  // lodash types broaden keys to string but this should be correct
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  keyBy(stateConfigs, "stateCode") as any;

/**
 * Exhaustive mapping of {@link StateConfig} objects for all supported states
 */
export const stateConfigsByUrlSlug: Record<
  string,
  StateConfig<StateCode> | undefined
> = keyBy(stateConfigs, "urlSlug");
