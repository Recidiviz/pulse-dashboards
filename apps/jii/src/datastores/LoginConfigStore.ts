// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { makeAutoObservable, set } from "mobx";

import { FlowMethod } from "~hydration-utils";

import { DataAPI } from "../apis/data/interface";
import {
  LandingPageConfig,
  StateCode,
  StateLandingPageConfig,
} from "../configs/types";

export type LoginConfigStoreExternals = { apiClient: DataAPI };

export class LoginConfigStore {
  constructor(private externals: LoginConfigStoreExternals) {
    makeAutoObservable(this);
  }

  landingPageConfig?: LandingPageConfig;

  stateLandingPageConfigs: Map<StateCode, StateLandingPageConfig> = new Map();

  *populateLandingPageConfig(): FlowMethod<DataAPI["landingPageConfig"], void> {
    this.landingPageConfig = yield this.externals.apiClient.landingPageConfig();
  }

  *populateStateLandingPageConfig(
    stateCode: StateCode,
  ): FlowMethod<DataAPI["stateLandingPageConfig"], void> {
    const config =
      yield this.externals.apiClient.stateLandingPageConfig(stateCode);
    set(this.stateLandingPageConfigs, { [stateCode]: config });
  }
}
