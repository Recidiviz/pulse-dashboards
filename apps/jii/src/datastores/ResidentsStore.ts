// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import keyBy from "lodash/keyBy";
import { makeAutoObservable, set } from "mobx";

import { ResidentRecord } from "~datatypes";
import { FilterParams } from "~firestore-api";
import { FlowMethod } from "~hydration-utils";

import { DataAPI } from "../apis/data/interface";
import { residentEligibilityReportConstructors } from "../configs/residentEligibilityReportConstructors";
import {
  IncarcerationOpportunityId,
  OpportunityRecord,
  ResidentsConfig,
} from "../configs/types";
import { EligibilityReport } from "../models/EligibilityReport/interface";
import type { RootStore } from "./RootStore";

// because we use a mapped type to ensure key-value agreement, this can't be an ES6 Map
type OpportunityRecordMapping = {
  [O in IncarcerationOpportunityId]?: OpportunityRecord<O>;
};

export class ResidentsStore {
  /**
   * Holds all resident records that have been fetched
   */
  residentsByExternalId: Map<string, ResidentRecord["output"]> = new Map();

  /**
   * Holds all opportunity eligibility records that have been fetched
   */
  residentOpportunityRecordsByExternalId: Map<
    string,
    OpportunityRecordMapping
  > = new Map();

  residentEligibilityReportsByExternalId: Map<
    string,
    Map<IncarcerationOpportunityId, EligibilityReport>
  > = new Map();

  constructor(
    private readonly rootStore: RootStore,
    public readonly config: ResidentsConfig,
  ) {
    makeAutoObservable(this);
  }

  private get apiClient() {
    return this.rootStore.apiClient;
  }

  get userStore() {
    return this.rootStore.userStore;
  }

  get opportunityIdsByUrlSlug(): Map<string, IncarcerationOpportunityId> {
    return new Map(
      Object.entries(this.config.incarcerationOpportunities).map(
        ([id, config]) => {
          return [
            config.urlSlug,
            id as keyof typeof this.config.incarcerationOpportunities,
          ];
        },
      ),
    );
  }

  /**
   * Get a convenient non-nullable ID value; accepts any string,
   * so you should be confident you have a valid slug or be prepared to handle an error.
   */
  opportunitySlugToIdOrThrow(slug: string) {
    const id = this.opportunityIdsByUrlSlug.get(slug);
    if (!id) {
      throw new Error(`No opportunity ID matches url segment ${slug}`);
    }
    return id;
  }

  areAllResidentsPopulated(): boolean {
    // if we started the session on a single resident's page, we might have had one
    // populated already. Seems a safe assumption that the total will always be > 1
    return this.residentsByExternalId.size > 1;
  }
  /**
   * Populates {@link residentsByExternalId} with the API response for all available residents.
   * Will not refetch if data is already populated, unless `forceRefresh` is true
   */
  *populateResidents(
    filters?: Array<FilterParams>,
    forceRefresh = false,
  ): FlowMethod<DataAPI["residents"], void> {
    if (!forceRefresh && this.areAllResidentsPopulated()) return;

    const residents = yield this.apiClient.residents(filters);
    set(this.residentsByExternalId, keyBy(residents, "personExternalId"));
  }

  isResidentPopulated(residentExternalId: string): boolean {
    return this.residentsByExternalId.has(residentExternalId);
  }

  /**
   * Populates {@link residentsByExternalId} with the API response for the resident
   * with personExternalId matching `residentExternalId`. Throws if the API request fails.
   * Will not refetch if data is already populated.
   */
  *populateResidentById(
    residentExternalId: string,
  ): FlowMethod<DataAPI["residentById"], void> {
    if (this.isResidentPopulated(residentExternalId)) return;

    const resident = yield this.apiClient.residentById(residentExternalId);
    this.residentsByExternalId.set(resident.personExternalId, resident);
  }

  get residentsByPseudoId(): Map<string, ResidentRecord["output"]> {
    return new Map(
      Object.entries(
        keyBy([...this.residentsByExternalId.values()], "pseudonymizedId"),
      ),
    );
  }

  isResidentWithPseudoIdPopulated(pseudoId: string): boolean {
    return this.residentsByPseudoId.has(pseudoId);
  }

  /**
   * Populates {@link residentsByExternalId} with the API response for the resident
   * with pseudonymizedId matching `residentPseudoId`. Throws if the API request fails.
   * Will not refetch if data is already populated.
   */
  *populateResidentByPseudoId(
    residentPseudoId: string,
  ): FlowMethod<DataAPI["residentById"], void> {
    if (this.isResidentWithPseudoIdPopulated(residentPseudoId)) return;

    const resident = yield this.apiClient.residentByPseudoId(residentPseudoId);
    this.residentsByExternalId.set(resident.personExternalId, resident);
  }

  isResidentEligibilityRecordPopulated(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): boolean {
    return (
      opportunityId in
      (this.residentOpportunityRecordsByExternalId.get(residentExternalId) ??
        {})
    );
  }

  /**
   * Populates {@link residentOpportunityRecordsByExternalId} with the API response for
   * the provided resident and opportunity IDs. Will populate with `undefined` if no record is found
   * (indicating the resident is ineligible). Will not refetch if data is already populated.
   */
  *populateOpportunityRecordByResidentId(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): FlowMethod<DataAPI["residentEligibility"], void> {
    if (
      this.isResidentEligibilityRecordPopulated(
        residentExternalId,
        opportunityId,
      )
    )
      return;

    const eligibilityRecord = yield this.apiClient.residentEligibility(
      residentExternalId,
      opportunityId,
    );
    const opportunityMap: OpportunityRecordMapping =
      this.residentOpportunityRecordsByExternalId.get(residentExternalId) ?? {};

    opportunityMap[opportunityId] = eligibilityRecord;

    if (!this.residentOpportunityRecordsByExternalId.has(residentExternalId)) {
      this.residentOpportunityRecordsByExternalId.set(
        residentExternalId,
        opportunityMap,
      );
    }
  }

  isResidentEligibilityReportPopulated(
    residentExternalId: string,
    opportunityId: IncarcerationOpportunityId,
  ): boolean {
    return !!this.residentEligibilityReportsByExternalId
      .get(residentExternalId)
      ?.has(opportunityId);
  }

  /**
   * Populates {@link residentEligibilityReportsByExternalId} with an {@link EligibilityReport} object
   * for the given resident and opportunity, if it does not already exist.
   */
  populateEligibilityReportFromData<OppId extends IncarcerationOpportunityId>(
    opportunityId: OppId,
    resident: ResidentRecord["output"],
    opportunity: OpportunityRecord<OppId> | undefined,
  ) {
    const residentExternalId = resident.personExternalId;

    if (
      this.isResidentEligibilityReportPopulated(
        residentExternalId,
        opportunityId,
      )
    ) {
      return;
    }

    const config = this.config.incarcerationOpportunities[opportunityId];

    if (!config) {
      throw new Error(`Opportunity ${opportunityId} is not configured`);
    }

    const ReportConstructor =
      residentEligibilityReportConstructors[opportunityId];

    const mapping =
      this.residentEligibilityReportsByExternalId.get(residentExternalId) ??
      new Map();

    mapping.set(
      opportunityId,
      new ReportConstructor(resident, config, opportunity),
    );
    if (!this.residentEligibilityReportsByExternalId.has(residentExternalId)) {
      this.residentEligibilityReportsByExternalId.set(
        residentExternalId,
        mapping,
      );
    }
  }

  get stateCode() {
    return this.userStore.stateCode;
  }
}
