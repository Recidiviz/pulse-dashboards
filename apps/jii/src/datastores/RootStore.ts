// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { captureException } from "@sentry/react";
import { makeObservable, onReactionError } from "mobx";

import { isOfflineMode, isTestEnv } from "~client-env-utils";
import { FlowMethod } from "~hydration-utils";

import { ApiClient } from "../apis/data/ApiClient";
import { DataAPI } from "../apis/data/interface";
import { OfflineAPIClient } from "../apis/data/OfflineAPIClient";
import { StateCode } from "../configs/types";
import { ResidentsStore } from "./ResidentsStore";
import { UiStore } from "./UiStore";
import { UserStore } from "./UserStore";

// global error handling for Mobx reactions
onReactionError((error) => {
  captureException(error);
});

export class RootStore {
  /**
   * Will be an offline client in offline and test modes,
   * and an external API client otherwise.
   */
  apiClient: DataAPI;

  residentsStore?: ResidentsStore;

  userStore: UserStore;

  uiStore: UiStore;

  // for convenience, this is a constant while we only have one state onboarded
  readonly stateCode: StateCode = "US_ME";

  constructor() {
    makeObservable(this, {
      populateResidentsStore: true,
      residentsStore: true,
    });

    this.userStore = new UserStore(this);

    this.uiStore = new UiStore();

    this.apiClient = this.createApiClient();
  }

  /**
   * Constructs and returns an appropriate API client based on the environment
   * (e.g. one that returns fixture data in offline mode). Not intended to be
   * MobX-reactive, since it creates a brand-new instance and discards any state
   * the previous instance might have held.
   */
  private createApiClient(): DataAPI {
    if (isOfflineMode() || isTestEnv()) {
      return new OfflineAPIClient(this);
    } else {
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      const store = this;
      const externals = {
        // these are getters rather than static properties so that the API client
        // can can observe changes to these values rather than getting a one-time snapshot
        get stateCode() {
          return store.stateCode;
        },
        get authClient() {
          return store.userStore.authClient;
        },
        get config() {
          return store.residentsStore?.config;
        },
      };

      return new ApiClient(externals);
    }
  }

  *populateResidentsStore(): FlowMethod<DataAPI["residentsConfig"], void> {
    const config = yield this.apiClient.residentsConfig();

    this.residentsStore = new ResidentsStore(this, config);
  }
}
