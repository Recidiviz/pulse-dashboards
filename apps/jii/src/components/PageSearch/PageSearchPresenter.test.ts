// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { waitFor } from "@testing-library/react";
import { keyBy } from "lodash";
import { set } from "mobx";

import { outputFixture, outputFixtureArray, usMeResidents } from "~datatypes";

import { residentsConfigByState } from "../../configs/residentsConfig";
import { ResidentsStore } from "../../datastores/ResidentsStore";
import { RootStore } from "../../datastores/RootStore";
import { PageSearchPresenter } from "./PageSearchPresenter";

let residentsStore: ResidentsStore;
let presenter: PageSearchPresenter;

beforeEach(() => {
  residentsStore = new ResidentsStore(
    new RootStore(),
    residentsConfigByState.US_ME,
  );
  presenter = new PageSearchPresenter(residentsStore);
});

describe("hydration", () => {
  test("needs hydration", () => {
    expect(presenter.hydrationState.status).toBe("needs hydration");
    expect(presenter.selectOptions).toEqual([]);
  });

  test("already hydrated", () => {
    set(
      residentsStore.residentsByExternalId,
      keyBy(outputFixtureArray(usMeResidents), "personExternalId"),
    );

    expect(presenter.hydrationState.status).toBe("hydrated");
  });

  test("hydrate", async () => {
    expect(presenter.hydrationState.status).toBe("needs hydration");

    presenter.hydrate();

    expect(presenter.hydrationState.status).toBe("loading");

    await waitFor(() =>
      expect(presenter.hydrationState.status).toBe("hydrated"),
    );

    expect(presenter.selectOptions).toMatchSnapshot();
  });
});

test("set active resident", async () => {
  await presenter.hydrate();

  expect(presenter.defaultOption).toBeUndefined();

  const expectedResident = outputFixture(usMeResidents[2]);

  presenter.setActiveResident(expectedResident.personExternalId);

  expect(presenter.defaultOption?.value).toBe(
    expectedResident.personExternalId,
  );
});
