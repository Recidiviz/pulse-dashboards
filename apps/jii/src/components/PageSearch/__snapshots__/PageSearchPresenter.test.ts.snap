// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hydration > hydrate 1`] = `
[
  {
    "label": "First Resident (RES001)",
    "value": "RES001",
  },
  {
    "label": "Second Resident (RES002)",
    "value": "RES002",
  },
  {
    "label": "Third Resident (RES003)",
    "value": "RES003",
  },
  {
    "label": "Fourth Resident (RES004)",
    "value": "RES004",
  },
  {
    "label": "Fifth Resident (RES005)",
    "value": "RES005",
  },
  {
    "label": "Sixth Resident (RES006)",
    "value": "RES006",
  },
  {
    "label": "Seventh Resident (RES007)",
    "value": "RES007",
  },
  {
    "label": "Eighth Resident (RES008)",
    "value": "RES008",
  },
  {
    "label": "Ineligible Resident (RES999)",
    "value": "RES999",
  },
]
`;
